### PR TITLE
Use more functional programming in ArenaMap::insert

### PR DESCRIPTION
**Pull request by kjeremy**
I find this more readable and it flattens out the body a little. Others may disagree.

### CODE REVIEW

**Comment by kjeremy** [comment]
I was wondering that myself. I'm not sure if `extend` does the reserve in once step. `resize` is another option but that could truncate.

**Comment by lnicola** [comment]
I think it can thanks to the size hint, but I'm not sure.

**Comment by kjeremy** [comment]
Ok I think it is. `Take` implements `TrustedLen` so the specializated gets used by SpecExtend.

**Comment by kjeremy** [comment]
I'm not sure what that failing macos test is about.

**Comment by lnicola** [comment]
I think it's a race, it happened in another PR or two today.

File: crates/ra_arena/src/map.rs [hunk 1]
```diff
@@ -14,14 +14,9 @@ pub struct ArenaMap<ID, V> {
...
-            self.v.reserve(idx + 1 - self.v.capacity());
-        }
-        if self.v.len() <= idx {
-            while self.v.len() <= idx {
-                self.v.push(None);
-            }
-        }
+
+        let fill = (idx + 1).saturating_sub(self.v.len());
+        self.v.extend(std::iter::repeat_with(|| None).take(fill));
```
Comment (matklad) [inline crates/ra_arena/src/map.rs]:
```suggestion
        self.v.resize(idx + 1, None);
```

Comment (matklad) [reply crates/ra_arena/src/map.rs]: Sorry for the imperative rain on your functional parade :)

Comment (matklad) [reply crates/ra_arena/src/map.rs]: ah, wait, it indeed can truncate, I guess that needs to be `(idx + 1).max(self.v.len())`

**Pull request merged**